crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -126,3 +168,160 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
     return 0.0;
 #endif
 }
+
+// NOTE: Due to the non-uniform control flow in `shadows::fetch_point_shadow`,
+// we must use the Level variant of textureSampleCompare to avoid undefined
+// behavior due to some of the fragments in a quad (2x2 fragments) being
+// processed not being sampled, and this messing with mip-mapping functionality.
+// The shadow maps have no mipmaps so Level just samples from LOD 0.
+fn sample_shadow_cubemap_hardware(light_local: vec3<f32>, depth: f32, light_id: u32) -> f32 {
+#ifdef NO_CUBE_ARRAY_TEXTURES_SUPPORT
+    return textureSampleCompare(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, depth);
+#else
+    return textureSampleCompareLevel(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, i32(light_id), depth);
+#endif
+}
+
+fn sample_shadow_cubemap_at_offset(
+    position: vec2<f32>,
+    coeff: f32,
+    x_basis: vec3<f32>,
+    y_basis: vec3<f32>,
+    light_local: vec3<f32>,
+    depth: f32,
+    light_id: u32,
+) -> f32 {
+    return sample_shadow_cubemap_hardware(
+        light_local + position.x * x_basis + position.y * y_basis,
+        depth,
+        light_id
+    ) * coeff;
+}
+
+// This more or less does what Castano13 does, but in 3D space. Castano13 is
+// essentially an optimized 2D Gaussian filter that takes advantage of the
+// bilinear filtering hardware to reduce the number of samples needed. This
+// trick doesn't apply to cubemaps, so we manually apply a Gaussian filter over
+// the standard 8xMSAA pattern instead.
+fn sample_shadow_cubemap_gaussian(
+    light_local: vec3<f32>,
+    depth: f32,
+    scale: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis so we can apply a 2D sampling pattern to a
+    // cubemap.
+    var up = vec3(0.0, 1.0, 0.0);
+    if (dot(up, normalize(light_local)) > 0.99) {
+        up = vec3(1.0, 0.0, 0.0);   // Avoid creating a degenerate basis.
+    }
+    let basis = orthonormalize(light_local, up) * scale * distance_to_light;
+
+    var sum: f32 = 0.0;
+    sum += sample_shadow_cubemap_at_offset(
+        D3D_SAMPLE_POINT_POSITIONS[0], D3D_SAMPLE_POINT_COEFFS[0],
+        basis[0], basis[1], light_local, depth, light_id);
```

Comment:
I hope that basis[2] is never used means its calculation is optimised away. :)

Reply:
Well, you can't optimize away its calculation, because it's used for the construction of the other two columns. As written, this is a dead store. However, this function will certainly be inlined, and the matrix will all be in GPU registers. So the register allocation liveness range analysis will trivially detect that it isn't used and returning the extra value will have no effect.

tl;dr this has no performance impact.